plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -21,54 +21,73 @@ class bp_connection_manager {
  public:
 #endif
 
-   static constexpr size_t max_bp_peers_per_producer = 4;
+   static constexpr size_t           max_bp_gossip_peers_per_producer = 8;
+   static constexpr fc::microseconds bp_gossip_peer_expiration = fc::hours(1);
+   static constexpr fc::microseconds my_bp_gossip_peer_expiration = fc::minutes(30); // resend my bp_peer info every 30 minutes
+   static constexpr fc::microseconds bp_gossip_peer_expiration_variance = bp_gossip_peer_expiration + fc::minutes(15);
+
    gossip_bp_index_t      gossip_bps;
 
+   struct bp_gossip_endpoint_t {
+      std::string server_endpoint;      // externally known inbound endpoint host:port
+      std::string outbound_ip_address;  // externally known outbound IP address
+
+      auto operator<=>(const bp_gossip_endpoint_t& rhs) const = default;
+      bool operator==(const bp_gossip_endpoint_t& rhs) const = default;
+   };
+
    // the following members are thread-safe, only modified during plugin startup
    struct config_t {
-      flat_map<account_name, net_utils::endpoint>   bp_peer_addresses;
-      flat_map<net_utils::endpoint, account_name>   bp_peer_accounts;
-      peer_name_set_t                               my_bp_accounts;       // block producer --producer-name
-      peer_name_set_t                               my_bp_peer_accounts;  // peer key account --p2p-producer-peer
+      // p2p-auto-bp-peer
+      flat_map<account_name, net_utils::endpoint>   auto_bp_addresses;      // --p2p-auto-bp-peer account->endpoint
+      flat_map<net_utils::endpoint, account_name>   auto_bp_accounts;       // --p2p-auto-bp-peer endpoint->account
+      // p2p-bp-gossip-endpoint, producer account -> [inbound_endpoint,outbound_ip_address] for bp gossip
+      flat_map<account_name, std::vector<bp_gossip_endpoint_t>>  my_bp_gossip_accounts;
```

Comment:
I'm worried about the `flat_map` containing a vector, and the insertion of new rows could cause possibly multiple vector copies (depending on the `flat_map` implementation).

`std::vector` can be trivially relocated to another memory position (just by copying the 3 pointers), but this relies on the [std::trivially_relocatable](https://isocpp.org/files/papers/P2786R13.html) trait which is not yet available, even in C++23.

It is possibly that boost's `flat_map` has its own trait for that, but I didn't see it with a cursory look. 

I think it would be safer in this case to use `std::map`, or even better `std::unordered_map` since I don't think ordered is a requirement here.

Reply:
https://github.com/AntelopeIO/spring/pull/1503/commits/07751bc9849efad8ac40e59ec425c37826477f71